server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java

```diff
@@ -194,44 +202,96 @@ private void handleGetOnUnpromotableShard(GetRequest request, IndexShard indexSh
+                observer.waitForNextChange(new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        getFromTranslog(request, indexShard, state, observer, l);
+                    }
+
+                    @Override
+                    public void onClusterServiceClose() {
+                        l.onFailure(new NodeClosedException(clusterService.localNode()));
+                    }
+
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        l.onFailure(new ElasticsearchException("Timed out retrying get_from_translog", cause));
+                    }
+                });
+            } else {
+                l.onFailure(e);
+            }
+        }));
+    }
+
+    private void tryGetFromTranslog(GetRequest request, IndexShard indexShard, ClusterState state, ActionListener<GetResponse> listener) {
+        ShardId shardId = indexShard.shardId();
+        var node = getCurrentNodeOfPrimary(state, shardId);
+        TransportGetFromTranslogAction.Request getFromTranslogRequest = new TransportGetFromTranslogAction.Request(request, shardId);
+        getFromTranslogRequest.setParentTask(request.getParentTask());
+        transportService.sendRequest(
+            node,
+            TransportGetFromTranslogAction.NAME,
+            getFromTranslogRequest,
+            new ActionListenerResponseHandler<>(listener.delegateFailure((l, r) -> {
+                if (r.getResult() != null) {
+                    logger.debug("received result for real-time get for id '{}' from promotable shard", request.id());
+                    l.onResponse(new GetResponse(r.getResult()));
+                } else {
+                    logger.debug(
+                        "no result for real-time get for id '{}' from promotable shard (segment generation to wait for: {})",
+                        request.id(),
+                        r.segmentGeneration()
+                    );
+                    if (r.segmentGeneration() == -1) {
+                        // Nothing to wait for (no previous unsafe generation), just handle the Get locally.
+                        ActionRunnable.supply(l, () -> shardOperation(request, shardId)).run();
+                    } else {
+                        assert r.segmentGeneration() > -1L;
+                        assert r.primaryTerm() > Engine.UNKNOWN_PRIMARY_TERM;
+                        indexShard.waitForPrimaryTermAndGeneration(
+                            r.primaryTerm(),
+                            r.segmentGeneration(),
+                            listener.delegateFailureAndWrap((ll, aLong) -> super.asyncShardOperation(request, shardId, ll))
+                        );
+                    }
+                }
+            }), TransportGetFromTranslogAction.Response::new, getExecutor(request, shardId))
```

Comment:
good catch. I'll follow this up in a new PR.

Reply:
@ywangd Checking this now, I don't think this is an issue. In TransportGetAction we also use `GET` threadpool in the constructor. So I don't think any change is necessary here. The `getExecutor` is enough.

Reply:
The `getExecutor` here ensures the response is handled with the right thread pool. It does nothing for request handling, i.e. `TransportGetFromTranslogAction` always handles the request with the `GET` thread pool. 

The `TransportGetAction` is indeed instantiated with `GET` thread pool. But it does _not_ just use it, it forks to the right thread pool with either [asyncGet](https://github.com/elastic/elasticsearch/blob/66a930ba460e90f233654383fd93c0698968fddc/server/src/main/java/org/elasticsearch/action/get/TransportGetAction.java#L181) or [asyncShardOperation](https://github.com/elastic/elasticsearch/blob/66a930ba460e90f233654383fd93c0698968fddc/server/src/main/java/org/elasticsearch/action/support/single/shard/TransportSingleShardAction.java#L111-L112). Before `handleGetOnUnpromotableShard` was added, `TransportGetAction` always forks. With `handleGetOnUnpromotableShard`, it sends transport actions without forking. This is fine since we have established that these are cheap enough to perform inline. My point is that the request handling on the otherside, in this case `TransportGetFromTranslogAction`, should fork to use the relevant thread pool which overall seems to match the behaviour before `handleGetOnUnpromotableShard` was introduced.